storage/src/index/storage.rs

```diff
@@ -2,359 +2,392 @@ use crate::index::Translator;
-            let prev = self.prev.take().unwrap();
-            unsafe {
-                (*prev_prev).next = (*prev).next.take();
-                self.next = (*prev_prev)
-                    .next
-                    .as_mut()
-                    .map(|next_next| next_next.as_mut() as *mut _);
-                self.prev = Some(prev_prev);
-            }
-            return;
-        }
-
-        let Some(occupied_entry) = self.entry.as_mut() else {
-            unreachable!("can_remove should prevent this");
-        };
 
-        // The element we are removing is at the front.
-        let head = occupied_entry.get_mut();
+        // Stop the iterator.
+        self.idx = 0;
+        self.last_idx = None;
+    }
 
-        match head.next.take() {
-            Some(next) => {
-                // There is a linked element, so just make it the new head.
-                *head = *next;
-                self.prev = None;
-                self.next = Some(head as *mut Record<V>);
-            }
-            None => {
-                // This is the only element, so removing it requires we remove the map entry
-                // entirely.
-                self.entry.take().unwrap().remove();
-                self.prev = None;
-                self.next = None;
-            }
+    /// Remove the last value returned by `next()` (swapping it with the most recently added value for the
+    /// translated key).
+    ///
+    /// This is a no-op if `next()` has not been called or `remove()` is called multiple times without
+    /// calling `next()` after each `remove()`.
+    pub fn remove(&mut self) {
+        let Some(i) = self.last_idx.take() else {
+            return;
+        };
+        self.values.swap_remove(i);
+        self.pruned.inc();
+        if i < self.idx {
+            self.idx -= 1;
         }
     }
 }
 
-impl<V> Record<V> {
-    fn iter(&self) -> ValueIterator<V> {
-        ValueIterator { next: Some(self) }
+impl<T: Translator, V> Drop for MutableIterator<'_, T, V> {
+    /// When the iterator is dropped, check if the `Vec` is empty (delete from the map) or
+    /// if it has only one value (demote to `One`).
+    fn drop(&mut self) {
+        unsafe {
+            let map = self.map.as_mut();
+            match self.values.len() {
+                0 => {
+                    // We have no values left, so remove the entry from the map.
+                    map.remove(&self.key);
+                }
+                1 => {
+                    // We have only one value left, so demote to `One`.
+                    let v = self.values.pop().unwrap();
+                    *map.get_mut(&self.key).unwrap() = Record::One(v);
+                }
+                _ => {
+                    // We have more than one value left, so do nothing.
+                }
+            }
+        }
     }
 }
 
-/// An index that maps translated keys to values.
+/// A memory-efficient index that maps arbitrary values to translated keys.
```

Comment:
Is the order switch here intentional? We are mapping keys to values, not the other way around.

Reply:
That was an auto-complete error when I was fixing up the comments. Great catch!